File: artifacts/tulipindicators/translate/src/msw.rs
use crate::*;
use std::f64::consts::PI;

pub fn ti_msw_start(options: Option<&[f64]>) -> i32 {
    // Check if options is None (equivalent to NULL check in C)
    // and if the slice has at least one element
    match options {
        Some(opt) if !opt.is_empty() => opt[0] as i32,
        _ => 0, // Default return value when options is None or empty
    }
}

pub fn helper_ti_msw_1(
    sine_idx_ref: &mut u32,
    lead_idx_ref: &mut u32,
    weight_ref: &mut f64,
    phase_ref: &mut f64,
    rp_ref: &mut f64,
    ip_ref: &mut f64,
    j_ref: &mut i32,
    input: &[f64],
    sine: &mut [f64],
    lead: &mut [f64],
    period: i32,
    pi: f64,
    tpi: f64,
    i: i32,
) {
    let mut sine_idx = *sine_idx_ref;
    let mut lead_idx = *lead_idx_ref;
    let mut weight = *weight_ref;
    let mut phase = *phase_ref;
    let mut rp = *rp_ref;
    let mut ip = *rp_ref;
    let mut j = *j_ref;

    rp = 0.0;
    ip = 0.0;

    for j in 0..period {
        weight = input[(i - j) as usize];
        rp += (f64::cos((tpi * j as f64) / period as f64) * weight);
        ip += (f64::sin((tpi * j as f64) / period as f64) * weight);
    }

    if f64::abs(rp) > 0.001 {
        phase = f64::atan(ip / rp);
    } else {
        phase = (tpi / 2.0) * if ip < 0.0 { -1.0 } else { 1.0 };
    }

    if rp < 0.0 {
        phase += pi;
    }

    phase += pi / 2.0;

    if phase < 0.0 {
        phase += tpi;
    }

    if phase > tpi {
        phase -= tpi;
    }

    sine[sine_idx as usize] = f64::sin(phase);
    sine_idx += 1;
    lead[lead_idx as usize] = f64::sin(phase + (pi / 4.0));
    lead_idx += 1;

    *sine_idx_ref = sine_idx;
    *lead_idx_ref = lead_idx;
    *weight_ref = weight;
    *phase_ref = phase;
    *rp_ref = rp;
    *ip_ref = ip;
    *j_ref = j;
}
pub fn ti_msw(
    size: i32,
    inputs: Option<&[&[f64]]>,
    options: Option<&[f64]>,
    outputs: Option<&mut [&mut [f64]]>,
) -> i32 {
    // Check for None options (equivalent to NULL checks in C)
    let options = match options {
        Some(opts) => opts,
        None => return 1,
    };
    
    let period = options[0] as i32;
    if period < 1 {
        return 1;
    }

    let start = ti_msw_start(Some(options));
    if size <= start {
        return 0;
    }

    // Unwrap inputs and outputs safely
    let input = match inputs {
        Some(inp) => inp[0],
        None => return 1,
    };
    
    let outputs = match outputs {
        Some(out) => out,
        None => return 1,
    };
    
    // Get the original outputs pointers before mutable borrow
    let outputs_ptr0 = outputs[0].as_ptr();
    let outputs_ptr1 = outputs[1].as_ptr();
    
    // Split the outputs into separate mutable slices
    let (sine_slice, rest) = outputs.split_at_mut(1);
    let sine = &mut sine_slice[0];
    let lead = &mut rest[0];
    
    let pi = PI;
    let tpi = 2.0 * pi;
    let mut weight = 0.0;
    let mut phase = 0.0;
    let mut rp = 0.0;
    let mut ip = 0.0;
    let mut j = 0;
    let mut sine_idx = 0;
    let mut lead_idx = 0;

    for i in period..size {
        helper_ti_msw_1(
            &mut sine_idx,
            &mut lead_idx,
            &mut weight,
            &mut phase,
            &mut rp,
            &mut ip,
            &mut j,
            input,
            sine,
            lead,
            period,
            pi,
            tpi,
            i,
        );
    }

    // Assertions translated to debug_assert! (only active in debug builds)
    debug_assert!((&sine[sine_idx as usize] as *const f64 as isize - outputs_ptr0 as isize) / std::mem::size_of::<f64>() as isize == (size - start) as isize);
    debug_assert!((&lead[lead_idx as usize] as *const f64 as isize - outputs_ptr1 as isize) / std::mem::size_of::<f64>() as isize == (size - start) as isize);

    0
}

File: artifacts/tulipindicators/decompose/src/msw.c
int ti_msw_start(const double *options)
{
  return (int) options[0];
}


/*** DEPENDENCIES:
***/


// hint:  ['sine_idx_ref is a mutable refrence to unsigned int', 'weight_ref is a mutable refrence to double', 'phase_ref is a mutable refrence to double', 'rp_ref is a mutable refrence to double', 'ip_ref is a mutable refrence to double', 'j_ref is a mutable refrence to int']
void helper_ti_msw_1(unsigned int * const sine_idx_ref, double * const weight_ref, double * const phase_ref, double * const rp_ref, double * const ip_ref, int * const j_ref, const double * const input, double * const sine, double * const lead, const int period, const double pi, const double tpi, int i)
{
  unsigned int sine_idx = *sine_idx_ref;
  double weight = *weight_ref;
  double phase = *phase_ref;
  double rp = *rp_ref;
  double ip = *ip_ref;
  int j = *j_ref;
  rp = 0;
  ip = 0;
  for (j = 0; j < period; j += 1)
  {
    weight = input[i - j];
    rp = rp + (cos((tpi * j) / period) * weight);
    ip = ip + (sin((tpi * j) / period) * weight);
  }

  if (fabs(rp) > .001)
  {
    phase = atan(ip / rp);
  }
  else
  {
    phase = (tpi / 2.0) * ((ip < 0) ? (-1.0) : (1.0));
  }
  if (rp < 0.0)
  {
    phase += pi;
  }
  phase += pi / 2.0;
  if (phase < 0.0)
  {
    phase += tpi;
  }
  if (phase > tpi)
  {
    phase -= tpi;
  }
  sine[sine_idx] = sin(phase);
  sine_idx += 1;
  *(lead++) = sin(phase + (pi / 4.0));
  *sine_idx_ref = sine_idx;
  *weight_ref = weight;
  *phase_ref = phase;
  *rp_ref = rp;
  *ip_ref = ip;
  *j_ref = j;
}


/*** DEPENDENCIES:
***/


int ti_msw(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *input = inputs[0];
  double *sine = outputs[0];
  unsigned int sine_idx = 0;
  double *lead = outputs[1];
  const int period = (int) options[0];
  if (period < 1)
  {
    return 1;
  }
  if (size <= ti_msw_start(options))
  {
    return 0;
  }
  const double pi = 3.1415926;
  const double tpi = 2 * pi;
  double weight = 0;
  double phase;
  double rp;
  double ip;
  int i;
  int j;
  for (i = period; i < size; i += 1)
  {
    helper_ti_msw_1(&sine_idx, &weight, &phase, &rp, &ip, &j, input, sine, lead, period, pi, tpi, i);
  }

  assert(((&sine[sine_idx]) - outputs[0]) == (size - ti_msw_start(options)));
  assert((lead - outputs[1]) == (size - ti_msw_start(options)));
  return 0;
}


/*** DEPENDENCIES:
int ti_msw_start(const double *options)
{
  return (int) options[0];
}


----------------------------
void helper_ti_msw_1(unsigned int * const sine_idx_ref, double * const weight_ref, double * const phase_ref, double * const rp_ref, double * const ip_ref, int * const j_ref, const double * const input, double * const sine, double * const lead, const int period, const double pi, const double tpi, int i)
{
  unsigned int sine_idx = *sine_idx_ref;
  double weight = *weight_ref;
  double phase = *phase_ref;
  double rp = *rp_ref;
  double ip = *ip_ref;
  int j = *j_ref;
  rp = 0;
  ip = 0;
  for (j = 0; j < period; j += 1)
  {
    weight = input[i - j];
    rp = rp + (cos((tpi * j) / period) * weight);
    ip = ip + (sin((tpi * j) / period) * weight);
  }

  if (fabs(rp) > .001)
  {
    phase = atan(ip / rp);
  }
  else
  {
    phase = (tpi / 2.0) * ((ip < 0) ? (-1.0) : (1.0));
  }
  if (rp < 0.0)
  {
    phase += pi;
  }
  phase += pi / 2.0;
  if (phase < 0.0)
  {
    phase += tpi;
  }
  if (phase > tpi)
  {
    phase -= tpi;
  }
  sine[sine_idx] = sin(phase);
  sine_idx += 1;
  *(lead++) = sin(phase + (pi / 4.0));
  *sine_idx_ref = sine_idx;
  *weight_ref = weight;
  *phase_ref = phase;
  *rp_ref = rp;
  *ip_ref = ip;
  *j_ref = j;
}


----------------------------
***/


